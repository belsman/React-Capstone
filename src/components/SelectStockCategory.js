import React from 'react';
import { useSelector, useDispatch } from 'react-redux';

const SelectStockCategory = () => {
  const dispatch = useDispatch();
  const currentSelectedFilter = useSelector(state => state.visibilityFilter);

  const handleInputChange = e => dispatch({ type: 'SET_VISIBILITY_FILTER', payload: e.target.value });

  return (
    <select name="stockCategory" id="stockCategory" value={currentSelectedFilter} onChange={handleInputChange}>
      <option value="activeStocks">Active Stocks</option>
      <option value="gainingStocks">Gainers</option>
      <option value="losingStocks">Losers</option>
    </select>
  );
};

export default SelectStockCategory;
