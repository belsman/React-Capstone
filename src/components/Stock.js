import React from 'react';
import { Link } from 'react-router-dom';
import PropTypes from 'prop-types';

const Stock = ({ stock }) => (
  <div className="card stock-card mb-1">
    <article className="card-body">
      <h5>{stock.ticker}</h5>
      <h6 className="text-truncate">{stock.companyName}</h6>
      <div className="my-2 py-2">
        <span className="">Price</span>
        <div className="d-flex justify-content-between">
          <span className="h5 text-secondary">
            $
            {' '}
            {stock.price}
          </span>
          <span className={stock.changes < 0 ? 'text-danger' : 'text-success'}>
            {stock.changesPercentage.replace(/[()]+/g, '')}
            {' '}
            <i className={stock.changes < 0 ? 'fa fa-arrow-down' : 'fa fa-arrow-up'} />
          </span>
        </div>
      </div>
      <Link to={`/stocks/${stock.ticker}`} className="btn btn-outline-info">
        View stock
      </Link>
    </article>
  </div>
);

Stock.propTypes = {
  stock: PropTypes.shape({
    ticker: PropTypes.string.isRequired,
    price: PropTypes.string.isRequired,
    changes: PropTypes.number.isRequired,
    changesPercentage: PropTypes.string.isRequired,
    companyName: PropTypes.string.isRequired,
  }).isRequired,
};

export default Stock;
