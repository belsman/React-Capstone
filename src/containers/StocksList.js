import { Link } from 'react-router-dom';
import { useSelector } from 'react-redux';

function StocksList() {
  const stocks = useSelector(state => state.stocks[state.visibilityFilter]);

  return (
    <section>
      {stocks.map(stock => (
        <article key={stock.ticker}>
          <h2>{stock.companyName}</h2>
          <h3>{stock.ticker}</h3>
          <ul>
            <li>{stock.price}</li>
            <li>{stock.changesPercentage}</li>
            <li>{stock.changes}</li>
          </ul>
          <Link to={`/stocks/${stock.ticker}`} className="button muted-button">
            View stock!
          </Link>
        </article>
      ))}
    </section>
  );
}

export default StocksList;
