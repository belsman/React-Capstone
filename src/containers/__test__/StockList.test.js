import React from 'react';
import { render, unmountComponentAtNode } from 'react-dom';
import { act } from 'react-dom/test-utils';
import pretty from 'pretty';

/* eslint-disable no-unused-vars */
import { useDispatch, useSelector } from 'react-redux';

import ClipLoader from 'react-spinners/ClipLoader';
import override from '../../syledComponent';

import StocksList from '../StocksList';
import Stock from '../../components/Stock';
import SelectStockCategory from '../../components/SelectStockCategory';

jest.mock('../../components/Stock', () => function Stock() {
  return <div>Hello</div>;
});

jest.mock('../../components/SelectStockCategory', () => function Stock() {
  return <div>Category</div>;
});

jest.mock('react-redux', () => ({
  useDispatch(fn) {
    return fn;
  },
  useSelector() {
    return {
      data: [{
        ticker: 'NVR',
        changes: -98.87,
        price: '3931.93',
        changesPercentage: '(-2.45%)',
        companyName: 'NVR Inc',
      }, {
        ticker: 'AMZN',
        changes: -73.38,
        price: '3294.62',
        changesPercentage: '(-2.18%)',
        companyName: 'Amazon.com Inc',
      }, {
        ticker: 'GOOG',
        changes: -50.8,
        price: '1591.04',
        changesPercentage: '(-3.09%)',
        companyName: 'Alphabet Inc.',
      },
      ],
      stockStatus: 'completed',
      stockError: null,
    };
  },
}));

jest.mock('../../syledComponent', () => function override() {
  return 'some css to';
});

jest.mock('react-spinners/ClipLoader', () => ({
  ClipLoader() {
    return (
      <div>Spin</div>
    );
  },
}));

let container = null;
beforeEach(() => {
  container = document.createElement('div');
  document.body.appendChild(container);
});

afterEach(() => {
  unmountComponentAtNode(container);
  container.remove();
  container = null;
});

it('renders with with the correct stocks-lists based on the store data', () => {
  act(() => {
    render(<StocksList />, container);
  });

  const stocks = document.querySelector('div.stock-list-container');
  expect(stocks.children).toHaveLength(3);
});

it('should render the expected output', () => {
  act(() => {
    render(<StocksList />, container);
  });

  expect(
    pretty(container.innerHTML),
  ).toMatchSnapshot();
});
