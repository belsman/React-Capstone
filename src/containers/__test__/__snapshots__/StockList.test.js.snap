// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render the expected output 1`] = `
"<div class=\\"border mb-2 p-2 d-sm-flex justify-content-sm-between\\"><small><a href=\\"https://financialmodelingprep.com/developer/docs/\\" class=\\"font-italic\\" target=\\"_blank\\" rel=\\"noreferrer\\">Data provided by Financial Modeling Prep</a></small>
  <div>Category</div>
</div>
<div class=\\"p-1\\">
  <div class=\\"stock-list-container\\">
    <div>Hello</div>
    <div>Hello</div>
    <div>Hello</div>
  </div>
</div>"
`;
